Handle errors during the archive/unarchive process

pr33thi opened the pull request:
[Fogbugz ticket](https://manage.dimagi.com/default.asp?283938#1543433)

**Expected behavior:**
- A case is closed
- The case close form is archived
- The case is rebuilt automatically, and the case is open again

**Current behavior when kafka is down:**
- A case is closed
- The case close form is archived
- Kafka is down, so the case is not rebuild automatically, and remains closed (but can be rebuilt manually)

**Solution:**
I created a context manager for the archive/unarchive functions (for both couch and sql), where a stub is created in the event of a failure somewhere in the un/archive process, and a celery task periodically re un/archives any forms with an associated stub.

**Things I am looking for feedback on**
- I have defined `date_queued` and `attempts` [here](https://github.com/dimagi/commcare-hq/compare/master...pv/archive-form-283938?expand=1#diff-6d9447ad465303860951a636dbf536aaR25). I only have them because they're included in the model for [a similar context manager](https://github.com/dimagi/commcare-hq/blob/master/corehq/ex-submodules/couchforms/migrations/0002_auto_20170802_1505.py). I think I can get rid of them but am not completely sure
- @calellowitz You mentioned a few edge cases earlier that I _think_ are covered here (ie I believe clicking on the archive button twice in a row is already handled in the archive/unarchive functions themselves), but unsure if I missed any
 
I tried to make the tests easy to follow, so I'd suggest looking at that first, then the rest of the code.

@calellowitz @dannyroberts 
code buddy @esoergel

pr33thi committed:
Create the unfinished archive stub class and sql mgiration

pr33thi committed:
Create the unfinished_archive context manager and use it in an archive function

pr33thi committed:
Add tests

pr33thi committed:
Create a periodic celery task that checks for unfinished archives, update tests accordingly

pr33thi committed:
Update the couch archive function as well

pr33thi committed:
Add the same logic and tests for unarchive

pr33thi committed:
Stickler

pr33thi committed:
Cleanup, reformat the tests to be more readable

snopoke commented on corehq/form_processor/submission_post.py:
```diff
@@ -541,3 +541,20 @@ def unfinished_submission(instance):
+
+
+@contextlib.contextmanager
+def unfinished_archive(instance, user_id, archive):
```
since this is unrelated to submissions consider moving to a different place (also fine to move the other context manager)

pr33thi replied:
Moved to a separate file: https://github.com/dimagi/commcare-hq/pull/22470/commits/dae93ea4d3ff1e0276e861accad83db7623eebe2

snopoke commented on corehq/form_processor/submission_post.py:
```diff
@@ -541,3 +541,20 @@ def unfinished_submission(instance):
+            archive=archive,
+            domain=instance.domain,
+        )
+    tracker = SubmissionProcessTracker(unfinished_archive_stub)
```
I don't think the use of the `tracker` is necessary here. You can just `yield` and then delete the stub afterwards.

snopoke replied:
I mean replace it with a plain yield and a delete:
```
yield 
unfinished_archive_stub.delete()
```

pr33thi replied:
https://github.com/dimagi/commcare-hq/pull/22470/commits/e5c7e9fd66f61ee296eb2f722622cdcf5a7910b1

snopoke commented on corehq/ex-submodules/couchforms/models.py:
```diff
@@ -346,27 +346,32 @@ def _meta_to_json(meta):
+        from corehq.form_processor.submission_post import unfinished_archive
+        with unfinished_archive(instance=self, user_id=user_id, archive=False):
+            self.doc_type = "XFormInstance"
+            self.history.append(XFormOperation(
```
Do we care if we end up with duplicate operations? Would be nice if we could avoid it.

pr33thi replied:
I changed it so that the history only gets updated at the end, after the form is archived and the case is rebuilt: https://github.com/dimagi/commcare-hq/pull/22470/commits/49d769180a6f9dc7fb3a779b14bb2181499ae2a1

One problem with this solution is that a form could be archived, but not have the archive show up in its history for the 5 minutes while it is being reprocessed.

On another branch, I worked on another solution where the history is only updated once, and not updated on retries, which has the reverse problem of showing that a form was archived before the case is rebuilt. https://github.com/dimagi/commcare-hq/commit/30fd6d86a3cdcc01adaf4fd388782b98e7cfc39a

I like the solution that I did in this branch better, it could lead to some confusion, but seems to target the root issue that was confusing to users in this ticket.

snopoke replied:
This won't work since rebuilding the case relies on the form being marked as 'archived' otherwise the case rebuild will have no effect. I think it's fine to do the alternate approach in https://github.com/dimagi/commcare-hq/commit/30fd6d86a3cdcc01adaf4fd388782b98e7cfc39a

pr33thi replied:
Hm, the tests do pass, which checks if the case is rebuilt properly. Are you referring to `is_archived`? Because that has to do with checking `self.doc_type`, which does get updated before the case is rebuilt

pr33thi replied:
Did the other approach here: https://github.com/dimagi/commcare-hq/pull/22470/commits/6e277f171ae513fec02b658970e11cf07bbda346

pr33thi replied:
https://github.com/dimagi/commcare-hq/pull/22470#issuecomment-441293835

snopoke reviewed (changes requested):
2 things:

1. Record length of 'queue' in datadog (number of stubs) + increment counter each time one is processed
2. Should make sure that all stubs are deleted prior to calling `archive` / `unarchive` to make sure that stub processing doesn't undo a user's action.

pr33thi committed:
Merge branch 'master' into pv/archive-form-283938

pr33thi committed:
Update the history after the message was successfully sent to kafka

pr33thi committed:
Record number of archive stubs in datadog

pr33thi committed:
Move context managers to separate file

pr33thi committed:
Change unfinished_archive_stub yield logic

pr33thi commented:
(FYI I pushed this too early, still in the middle of addressing comments)

pr33thi committed:
Only update datadog at the beginning of each celery task

pr33thi committed:
Delete stubs before any archive operation initiated by a user

pr33thi committed:
Make user_id field nullable so this doesn't change the behavior in unexpected ways

pr33thi commented:
@snopoke addressed your comments:
1. Record length of 'queue' in datadog (number of stubs) + increment counter each time one is processed: https://github.com/dimagi/commcare-hq/pull/22470/commits/02ac56b6b8f5a1756e7b1d3b2415783cba4c5616 and then to address your comment: https://github.com/dimagi/commcare-hq/pull/22470/commits/867f9e93506365ae34fd704e3d840975fc3d4e01
2. Should make sure that all stubs are deleted prior to calling archive / unarchive to make sure that stub processing doesn't undo a user's action: https://github.com/dimagi/commcare-hq/pull/22470/commits/eb34144c5b84204d53910fe44c605c6497f8c010

pr33thi committed:
Exit this task after a set time period

snopoke commented on corehq/apps/userreports/tasks.py:
```diff
@@ -238,9 +239,27 @@ def delete_data_source_task(domain, config_id):
+@periodic_task(serializer='pickle', run_every=crontab(minute='*/5'), queue=settings.CELERY_PERIODIC_QUEUE)
+def reprocess_archive_stubs():
+    # Exit this task after 1 minute so that the same stub isn't ever processed in multiple queues.
+    check_repeaters()
```
I think there was a misunderstanding here. This was an example of a task that exits after a certain time period. Calling it here won't cause this task to exit and it will also fire off all the repeaters scheduled to run. You need to do something like this:
```python
start = time.time()
for .....:
    if time.time() - start > cutoff:
        return
```

pr33thi replied:
My mistake: https://github.com/dimagi/commcare-hq/pull/22470/commits/0ae907560cff5e22e900a6146f2c5c930433cab2

pr33thi committed:
Cut off task after 4 minutes

pr33thi committed:
Alternate approach for avoiding duplicate history entries: Only update the history on user initiated actions

pr33thi committed:
Update model to have history_added, and make a tracker for the context manager

pr33thi committed:
Modify couch un/archive logic to set history_updated, have a function that only sends to kafka

pr33thi committed:
Update tests

pr33thi committed:
Modify sql un/archive logic to set history_updated, have a function that only sends to kafka

pr33thi committed:
Update the celery task to only call the send to kafka if that is the only thing that hasnt been run yet

pr33thi committed:
Add a sql test mock

pr33thi committed:
Update test_form_dbaccessor to use context manager

pr33thi committed:
Add absolute import

pr33thi commented:
Implemented changes that @snopoke suggested on slack:

Simon Kelly [9:22 AM]
re that commit, I think it's a bit weird to skip only the one sql call during retry
I understand why you're skipping it but if you're skipping one you should skip both  (not that I necessarily think that's a good idea either since if there was an error during the SQL call you'd want to re-do both)
you could the same pattern as for form submissions:
```with stub:
    mark_form_archived_and_create_history()
    stub.form_archived = True
    stub.save()
    xform_archived.send(...)```

then on retry you can do this:

```if stub.form_archived:
    # just call the signal
    xform_archived.send(..)
else:
    # redo the whole process
    xform.archive(...)
```

pr33thi commented on corehq/form_processor/submission_post.py:
```diff
@@ -511,33 +510,3 @@ def notify_submission_error(instance, exception, message):
         logging.error(message, exc_info=sys.exc_info(), extra={'details': details})
-
-
-class SubmissionProcessTracker(object):
```
Note, I deleted this from here and moved it to `corehq/form_processor/submission_process_tracker.py`.

snopoke commented on corehq/ex-submodules/couchforms/models.py:
```diff
@@ -346,27 +346,54 @@ def _meta_to_json(meta):
-        ))
-        self.save()
-        xform_archived.send(sender="couchforms", xform=self)
+    def archive(self, user_id=None, retry_archive=False):
```
I think you can get rid of the `retry_archive` kwarg and related branching

pr33thi replied:
I think it should stay for the case where the doctype successfully updated but nothing else went through. If that's the case, then on retry (for an archive), it'll see that `self.is_archived` is True and it'll skip the rest of the logic.

snopoke replied:
I think we're missing each other a bit here. If we're updating the stub after the doc_type and history are saved then we know that they have been updated:
```
with stub:
    form.doc_type = 'X'
    form.save()
    stub.archive_history_updated()
    signal.send(...)
```
I noticed below that you were calling `stub.archive_history_updated()` before the form was saved.

snopoke commented on corehq/ex-submodules/couchforms/models.py:
```diff
@@ -517,3 +544,28 @@ def __unicode__(self):
+    history_updated = models.BooleanField(default=False)
+    domain = models.CharField(max_length=256)
+    date_queued = models.DateTimeField(null=True, db_index=True)
+    attempts = models.IntegerField(default=0)
```
don't think `date_queued` or `attempts` are used

pr33thi replied:
https://github.com/dimagi/commcare-hq/pull/22470/commits/149e758f09abaf32028c30f1f33145bb9268e55b

snopoke commented on corehq/form_processor/backends/sql/dbaccessors.py:
```diff
@@ -609,16 +611,22 @@ def soft_delete_forms(domain, form_ids, deletion_date=None, deletion_id=None):
+    def _archive_unarchive_form(self, form, user_id, archive, archive_stub):
+        form_id = form.form_id
+        with get_cursor(XFormInstanceSQL) as cursor:
+            cursor.execute('SELECT archive_unarchive_form(%s, %s, %s)', [form_id, user_id, archive])
```
why did this need to get split out form the other SQL query?

pr33thi replied:
Because the logic in `_archive_unarchive_form` updates the history (but I can make the name of the function more clear) and `send_to_kafka` sends it to kafka (but maybe @calellowitz can double check that for me?). I was trying to do the same separation of logic here in sql that I put into the couch logic.

snopoke replied:
I don't think it's necessary to separate these if we are marking the stub as 'updated' after these have been called.

snopoke commented on corehq/form_processor/backends/sql/dbaccessors.py:
```diff
@@ -609,16 +611,22 @@ def soft_delete_forms(domain, form_ids, deletion_date=None, deletion_id=None):
+        form_id = form.form_id
+        with get_cursor(XFormInstanceSQL) as cursor:
+            cursor.execute('SELECT archive_unarchive_form(%s, %s, %s)', [form_id, user_id, archive])
+            archive_stub.history_updated = True
```
is this supposed to be `stub.archive_history_updated()`?

pr33thi replied:
Oops, https://github.com/dimagi/commcare-hq/pull/22470/commits/3a78517e70a3b506f12977c2111304964431b2ab

snopoke commented on corehq/form_processor/submission_process_tracker.py:
```diff
@@ -0,0 +1,69 @@
+    def __init__(self, stub=None):
+        self.stub = stub
+
+    def delete_me_get_history(self):
```
what's this for?

pr33thi replied:
Oops, did not mean to check it in: https://github.com/dimagi/commcare-hq/pull/22470/commits/75b23b43fb7b8aafe388c51f447ebb2ba9df7e04

pr33thi committed:
Clarify some of the comments

pr33thi committed:
Remove unused fields from UnfinishedArchiveStub model

pr33thi committed:
Clarify function name to include the word history

pr33thi committed:
Use function to update archive history

pr33thi committed:
Delete unused function

pr33thi commented:
FYI @snopoke this is ready for review again!

pr33thi committed:
Fix the test mock

pr33thi committed:
Merge branch 'master' into pv/archive-form-283938

pr33thi committed:
Remove get_attachments_for_forms, which was accidentally added back during merge

pr33thi commented on corehq/form_processor/backends/sql/dbaccessors.py:
```diff
@@ -519,21 +519,6 @@ def hard_delete_forms(domain, form_ids, delete_attachments=True):
         return deleted_count
 
-    @staticmethod
-    def get_attachments_for_forms(form_ids, ordered=False):
```
This was deleted here: https://github.com/dimagi/commcare-hq/commit/361d336e09101cf80d6ba8e0ee0d4caea1084ea9#diff-d28c016fb16e34a86ad9ce844e326789L535
It must have gotten added back when I merged with master, so I deleted it here.

pr33thi commented on corehq/form_processor/backends/sql/dbaccessors.py:
```diff
@@ -520,16 +520,33 @@ def hard_delete_forms(domain, form_ids, delete_attachments=True):
+        assert isinstance(form_ids, list)
+        if not form_ids:
+            return []
+        attachments = list(XFormAttachmentSQL.objects.raw(
```
https://github.com/dimagi/commcare-hq/pull/22470/commits/52a9134af8f2be2d89ab247ea7389596e0989a02

snopoke commented on corehq/ex-submodules/couchforms/models.py:
```diff
@@ -346,27 +346,56 @@ def _meta_to_json(meta):
+                user=user_id,
+                operation='archive',
+            ))
+            archive_stub.archive_history_updated()
```
This should be after the save

pr33thi replied:
https://github.com/dimagi/commcare-hq/pull/22470/commits/a2a9850d9000a53e1294aee61d649dd23ca55d7f

snopoke commented on corehq/ex-submodules/couchforms/models.py:
```diff
@@ -346,27 +346,56 @@ def _meta_to_json(meta):
+                user=user_id,
+                operation='unarchive',
+            ))
+            archive_stub.archive_history_updated()
```
should be after the save

pr33thi replied:
https://github.com/dimagi/commcare-hq/pull/22470/commits/a2a9850d9000a53e1294aee61d649dd23ca55d7f

snopoke commented on corehq/ex-submodules/couchforms/models.py:
```diff
@@ -346,27 +346,56 @@ def _meta_to_json(meta):
+            # If this archive was initiated by a user, delete all other stubs for this action so that this action
+            # isn't overridden
+            from couchforms.models import UnfinishedArchiveStub
+            UnfinishedArchiveStub.objects.filter(user_id=user_id).all().delete()
```
I think you only need to delete the ones for this form right?

pr33thi replied:
Replaced user_id with form_id here: https://github.com/dimagi/commcare-hq/pull/22470/commits/9a3857e686b4c3d205e61e1c3b12a2e77bd09a1e

snopoke commented on corehq/form_processor/models.py:
```diff
@@ -551,20 +551,51 @@ def get_xml(self):
+            archive_stub.archive_history_updated()
+            xform_unarchived.send(sender="form_processor", xform=self)
+
+    def send_archive_to_kafka(self, user_id):
```
I don't think these functions are necessary, you can just call `xform_archived.send(sender=sender, xform=self)` in the task (assuming we agree on previous comments)

snopoke commented on corehq/ex-submodules/couchforms/models.py:
```diff
@@ -346,27 +346,56 @@ def _meta_to_json(meta):
-        ))
-        XFormInstance.save(self)  # subclasses explicitly set the doc type so force regular save
-        xform_unarchived.send(sender="couchforms", xform=self)
+    def send_unarchive_to_kafka(self, user_id):
```
can just call `xform_unarchived.send(sender="couchforms", xform=self)` directly from the task

snopoke reviewed (commented):
@pr33thi seems we're missing each other a bit here. Let me know if you want to go to voice.

pr33thi committed:
Say stub history was updated after save()

pr33thi committed:
Merge branch 'master' into pv/archive-form-283938

pr33thi committed:
Filter stubs on xform id, not user id

pr33thi commented:
Will address the remaining 3 comments after discussion with @snopoke 

https://github.com/dimagi/commcare-hq/pull/22470#discussion_r237844843

https://github.com/dimagi/commcare-hq/pull/22470#discussion_r237845276

https://github.com/dimagi/commcare-hq/pull/22470#discussion_r237845355

pr33thi committed:
Fix bugs

pr33thi committed:
Get rid of retry_archive logic

pr33thi committed:
Simplify the send_archive_action_to_kafka logic

pr33thi committed:
Only delete the stub in the functions in models.py

snopoke commented on corehq/ex-submodules/couchforms/models.py:
```diff
@@ -349,24 +349,47 @@ def xml_md5(self):
+        # Delete the original stub
+        UnfinishedArchiveStub.objects.filter(xform_id=self.form_id).all().delete()
+        with unfinished_archive(instance=self, user_id=user_id, archive=archive):
+            xform_archived.send(sender="couchforms", xform=self)
```
this should be `xform_unarchived` when `archive == False`

pr33thi replied:
https://github.com/dimagi/commcare-hq/pull/22470/commits/d393107933168eeeb3f25f3319cb8d25e25bdbad

snopoke commented on corehq/form_processor/backends/sql/dbaccessors.py:
```diff
@@ -584,16 +584,16 @@ def soft_delete_forms(domain, form_ids, deletion_date=None, deletion_id=None):
-    @staticmethod
     @transaction.atomic
-    def _archive_unarchive_form(form, user_id, archive):
+    def send_to_kafka(self, form, user_id, archive):
```
should undo this rename since the previous name is more correct

pr33thi replied:
https://github.com/dimagi/commcare-hq/pull/22470/commits/41f1f3a3932c312fe24b6bda81c9e8a7be828e10

snopoke commented on corehq/form_processor/models.py:
```diff
@@ -552,19 +552,43 @@ def xml_md5(self):
+        # If this archive was initiated by a user, delete all other stubs for this action so that this action
+        # isn't overridden
+        from couchforms.models import UnfinishedArchiveStub
+        UnfinishedArchiveStub.objects.filter(xform_id=self.form_id).all().delete()
```
I think this should happen after the check for `self.is_archived` as you have it in the couch version. Same below.

pr33thi replied:
https://github.com/dimagi/commcare-hq/pull/22470/commits/2c517d66b377e2c5d2147d56f4fa9c20d5027558

snopoke commented on corehq/form_processor/models.py:
```diff
@@ -552,19 +552,43 @@ def xml_md5(self):
+        # Delete the original stub
+        UnfinishedArchiveStub.objects.filter(xform_id=self.form_id).all().delete()
+        with unfinished_archive(instance=self, user_id=user_id, archive=archive):
+            xform_archived.send(sender="form_processor", xform=self)
```
should be `xform_unarchived` when `archived == False`

pr33thi replied:
https://github.com/dimagi/commcare-hq/pull/22470/commits/d393107933168eeeb3f25f3319cb8d25e25bdbad

snopoke reviewed (commented):
Almost there!

pr33thi committed:
Call xform_unarchived or xform_archived.send at the right times

pr33thi committed:
Change function name back

pr33thi committed:
Change order on sql archive to match couch archive

snopoke commented on corehq/apps/userreports/tasks.py:
```diff
@@ -238,9 +239,35 @@ def delete_data_source_task(domain, config_id):
+            if stub.archive:
+                xform.publish_archive_action_to_kafka(user_id=stub.user_id, archive=True)
+            else:
+                xform.publish_archive_action_to_kafka(user_id=stub.user_id, archive=False)
```
nit: just noticed this now but this could be one line:
```
 xform.publish_archive_action_to_kafka(user_id=stub.user_id, archive=stub.archive)
```

pr33thi replied:
good point: https://github.com/dimagi/commcare-hq/pull/22470/commits/c629897b1c835844b226f82dd4a3e34c09e6330b

snopoke commented on corehq/form_processor/backends/sql/dbaccessors.py:
```diff
@@ -584,16 +584,16 @@ def soft_delete_forms(domain, form_ids, deletion_date=None, deletion_id=None):
-    @staticmethod
     @transaction.atomic
-    def _archive_unarchive_form(form, user_id, archive):
+    def _archive_unarchive_form(self, form, user_id, archive):
```
this can be changed back to static

pr33thi replied:
https://github.com/dimagi/commcare-hq/pull/22470/commits/543f0c30a88142f1452ba2676affc6cca8aa2ddd

snopoke commented on corehq/form_processor/backends/sql/dbaccessors.py:
```diff
@@ -520,16 +520,16 @@ def hard_delete_forms(domain, form_ids, delete_attachments=True):
+    def unarchive_form(form, user_id):
         from corehq.form_processor.change_publishers import publish_form_saved
-        FormAccessorSQL._archive_unarchive_form(form, user_id, False)
+        FormAccessorSQL()._archive_unarchive_form(form, user_id, archive=False)
```
can remove the `()` after the method is changed back to static (same above)

pr33thi replied:
https://github.com/dimagi/commcare-hq/pull/22470/commits/543f0c30a88142f1452ba2676affc6cca8aa2ddd

snopoke reviewed (approved):
small comments but not blocking

pr33thi committed:
Consolidate logic

pr33thi committed:
Change function to static

dannyroberts commented:
Just saws this was merged. Have we had the opportunity to test this in a production-like environment yet? If not I'd advocate for throwing this on staging and testing basic things like (1) submit a form (2) find form in submit history (3) click through to the form detail page (4) on that page, archive it (5) make sure it's actually archived (5) unarchive it (6) make sure it's actually unarchived.

Otherwise we're opening ourselves up to the possibility deploying something in a really core part of our system that has bugs we could have easily caught. Apologies if I just missed the place that this conversation happened.

pr33thi commented:
@dannyroberts @snopoke 
I reverted this PR just in case, pushed the changes to staging, and successfully did all the steps that Danny listed. Does anyone have any more suggestions on things to test out, or is that + the tests passing successfully adequate for me to merge this back in?

snopoke commented:
I think that's fine. This doesn't affect any core form processing, only arching and unarchiving which is much less common. If you've tested what Danny suggested then I think it's fine to proceed.

dannyroberts commented:
@pr33thi Thanks for humoring me with the sanity check on staging. I think you're all set to revert the revert! (Feel free to merge it yourself and reference this PR/comment on that revert-revert PR.)
